lib/class-wp-theme-json.php

```diff
@@ -1003,6 +1003,47 @@ private function get_block_styles() {
 				);
 			}
 
+			// Some properties can be styled by the user
+			// and will end up having preset classes attached to the block
+			// such as `.has-text-color .has-red-color`, etc.
+			//
+			// In those cases, we want the user preferences
+			// to overrule the theme's, hence this negations.
```

Comment:
This sounds very fragile to me and not exhaustive, what do you mean by "user preferences" here?

Reply:
The issue here is that preset classes are enqueued in the global stylesheet as:

`.has-red-color { color: red; }`

while the post title block (and potentially others now that block supports can be used to target inner elements) can do:

`h2.wp-block-post-title { color: "theme-color"; } `

Hence, the theme setting specificity won't let the user preference to take effect (user selects a preset color for the block => the class gets attached to the post content).

Alternatives I've considered:

1. Bumping up the specificity of the preset classes => not robust, we can't be sure what's the higher block specificity we have to counteract.

2. Add more block selectors to the preset classes rulesets, so we'd end up with:

```css
h2.wp-block-post-title.has-red-color,
<other-selectors>,
.has-red-color { 
  color: red
}
```

To avoid having to add all block selectors for all preset classes, which is a lot of CSS, I think we should output only the selectors for blocks whose corresponding property is styled (if they have a color => add selector to list).

3. Finally, I thought about the current approach: check for the presence of `has-text-color`, etc. It outputs a lot less CSS. I think it addresses all the cases we have. The only thing I'm not 100% comfortable with is the font-size check, as we only output `.has-<value>-font-size` in those cases so the attribute selector can only check for `-font-size`. Theoretically, I understand this bit may report false positives, but I'm not sure it'll be an issue in practice.

I have a slight preference for 3 given the amount of CSS we remove from shipping, but 2 would also work.

Reply:
Don't you think when a preset is applied, it should always be favored? in which case, adding `important` (forget the dogma) to the generated styles could be actually the best way to go here?

Reply:
I'm saying that because I don't believe relying on `has-text-color` and similar classes is a good thing, it adds an indirect dependency to something the block might or might not apply. (the block supports should be seen as just a shortcut for block authors and not something mandatory when using presets...)

Reply:
Ah, I see your point now: it's about the fact that `has-text-color` classes are only added by _core's UI controls_ and not necessarily by any other custom UI controls. :ok_hand:

Actually. Given how the style system works, `!important` is a good choice for preset classes the user attaches to the post content. If those values need to be overridden, in addition to regular CSS, we can offer hooks for the theme.json processing, so plugins have an alternate mechanism to modify this behavior.

Thanks for this feedback!

Reply:
https://github.com/WordPress/gutenberg/pull/29533/commits/980a0b6c981970388430d374cbe5d017c865c29c